langfuse-core/src/index.ts

```diff
@@ -674,14 +687,207 @@ abstract class LangfuseCoreStateless {
+          const media = new LangfuseMedia({
+            base64DataUri: `data:audio/${data.audio["format"] || "wav"};base64,${data.audio.data}`,
+          });
+
+          await this.processMediaItem({ media, traceId, observationId, field });
+
+          return {
+            ...data,
+            audio: {
+              ...data.audio,
+              data: media,
+            },
+          };
+        }
+
+        // Recursively process nested objects
+        return Object.fromEntries(
+          await Promise.all(
+            Object.entries(data).map(async ([key, value]) => [key, await processRecursively(value, level + 1)])
+          )
+        );
+      }
+
+      return data;
+    };
+
+    return await processRecursively(data, 1);
+  }
+
+  private async processMediaItem({
+    media,
+    traceId,
+    observationId,
+    field,
+  }: {
+    media: LangfuseMedia;
+    traceId: string;
+    observationId?: string;
+    field: string;
+  }): Promise<void> {
+    try {
+      if (!media.contentLength || !media._contentType || !media.contentSha256Hash || !media._contentBytes) {
+        return;
+      }
+
+      const getUploadUrlBody: GetMediaUploadUrlRequest = {
+        contentLength: media.contentLength,
+        traceId,
+        observationId,
+        field,
+        contentType: media._contentType,
+        sha256Hash: media.contentSha256Hash,
+      };
+
+      const fetchResponse = await this.fetch(
+        `${this.baseUrl}/api/public/media`,
+        this._getFetchOptions({
+          method: "POST",
+          body: JSON.stringify(getUploadUrlBody),
+        })
+      );
+
+      const uploadUrlResponse = (await fetchResponse.json()) as GetMediaUploadUrlResponse;
+
+      const { uploadUrl, mediaId } = uploadUrlResponse;
+      media._mediaId = mediaId;
+
+      if (uploadUrl) {
+        this._events.emit("debug", `Uploading media ${mediaId}`);
+        const startTime = Date.now();
+        const uploadResponse = await this.fetch(uploadUrl, {
+          method: "PUT",
+          body: media._contentBytes,
+          headers: {
+            "Content-Type": media._contentType,
+            "x-amz-checksum-sha256": media.contentSha256Hash,
+          },
+        });
+
+        const patchMediaBody: PatchMediaBody = {
+          uploadedAt: new Date().toISOString(),
+          uploadHttpStatus: uploadResponse.status,
+          uploadHttpError: await uploadResponse.text(),
+          uploadTimeMs: Date.now() - startTime,
+        };
+
+        await this.fetch(
+          `${this.baseUrl}/api/public/media/${mediaId}`,
+          this._getFetchOptions({ method: "PATCH", body: JSON.stringify(patchMediaBody) })
+        );
+        this._events.emit("debug", `Media upload status reported for ${mediaId}`);
+      } else {
+        this._events.emit("debug", `Media ${mediaId} already uploaded`);
+      }
+    } catch (err) {
+      this._events.emit("error", `Error processing media item: ${err}`);
+    }
+  }
```

Comment:
for all fetch() calls in this function, what happens on errors / wrong status codes?

I assume that uploadUrlResponse in `const uploadUrlResponse = (await fetchResponse.json()) as GetMediaUploadUrlResponse;` will for example contain wrong data.

Reply:
The fetch api throws only on network errors (timeout, connection failure, dns lookup...) but not if status code is 4xx / 5xx . It exposes a `ok` method to check whether range is 2xx.

I have not yet implemented retries, that will be a follow up task as I have to budget time